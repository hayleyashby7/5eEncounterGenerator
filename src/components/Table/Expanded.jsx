/* eslint-disable react/forbid-prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import Monster from '../Monster/Monster';

function Expanded({ data, colSpan }) {
    return (
        <tr>
            <td colSpan={colSpan}>
                <Monster data={data} />
            </td>
        </tr>
    );
}

Expanded.propTypes = {
    data: PropTypes.object.isRequired,
    colSpan: PropTypes.number.isRequired,
};

export default Expanded;
