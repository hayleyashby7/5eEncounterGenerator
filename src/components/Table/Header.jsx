import React from 'react';
import PropTypes from 'prop-types';

function Header({ header }) {
    return (
        <th key={header} className={header.important ? 'table-cell' : 'hidden md:table-cell'}>
            {header.name}
        </th>
    );
}

Header.propTypes = {
    header: PropTypes.shape({
        name: PropTypes.string.isRequired,
        important: PropTypes.bool.isRequired,
    }).isRequired,
};

export default Header;
