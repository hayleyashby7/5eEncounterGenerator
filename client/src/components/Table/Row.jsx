/* eslint-disable import/no-extraneous-dependencies */
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlus, faMinus } from '@fortawesome/free-solid-svg-icons';
import Expanded from './Expanded';

function Row({ data, cols }) {
    const [isExpanded, setIsExpanded] = useState(false);

    const handleExpand = () => {
        setIsExpanded(!isExpanded);
    };

    return (
        <>
            <tr>
                <td className='table-cell'>{data.name}</td>
                <td className='table-cell'>{data.challenge_rating}</td>
                <td className='hidden md:table-cell'>{data.type}</td>
                <td className='hidden md:table-cell'>{data.size}</td>
                <td className='hidden md:table-cell'>{data.alignment}</td>
                <td>
                    <div className='table-cell '>
                        <FontAwesomeIcon
                            icon={isExpanded ? faMinus : faPlus}
                            onClick={handleExpand}
                            title='Expand'
                        />
                    </div>
                </td>
            </tr>
            {isExpanded && <Expanded data={data} colSpan={cols} />}
        </>
    );
}

Row.propTypes = {
    data: PropTypes.shape({
        name: PropTypes.string.isRequired,
        challenge_rating: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired,
        size: PropTypes.string.isRequired,
        alignment: PropTypes.string.isRequired,
    }).isRequired,
    cols: PropTypes.number.isRequired,
};

export default Row;
