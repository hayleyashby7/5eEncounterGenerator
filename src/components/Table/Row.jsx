import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Expanded from './Expanded';

function Row({ data }) {
    const [isExpanded, setIsExpanded] = useState(false);

    const handleExpand = () => {
        setIsExpanded(!isExpanded);
    };

    return (
        <>
            <tr onClick={handleExpand}>
                <td>{data.name}</td>
                <td>{data.challenge_rating}</td>
                <td>{data.type}</td>
                <td>{data.size}</td>
                <td>{data.alignment}</td>
            </tr>
            {isExpanded ? <Expanded /> : null}
        </>
    );
}

Row.propTypes = {
    data: PropTypes.shape({
        name: PropTypes.string.isRequired,
        challenge_rating: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired,
        size: PropTypes.string.isRequired,
        alignment: PropTypes.string.isRequired,
    }).isRequired,
};

export default Row;
