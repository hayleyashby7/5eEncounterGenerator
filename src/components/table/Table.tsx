import Header, {HeaderType} from './Header';

interface TableProps {
    headers: HeaderType[];
  
}

export const Table: React.FC<TableProps> = ({ headers }) => {
    return (
        <table className='table w-full border-separate text-left'>
            <thead>
                <tr>
                    {headers.map((header) => (
                        <Header key={header.name} header={header} />
                    ))}
                </tr>
            </thead>
            <tbody>
                { }
            </tbody>
        </table>
    );
};

export default Table;