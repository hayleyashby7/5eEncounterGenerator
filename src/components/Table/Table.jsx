/* eslint-disable react/forbid-prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import Row from './Row';

function Table({ headers, data }) {
    return (
        <table>
            <thead>
                <tr>
                    {headers.map((header) => (
                        <th key={header}>{header}</th>
                    ))}
                </tr>
            </thead>
            <tbody>
                {data.map((item) => (
                    <Row key={item.slug} data={item} cols={headers.length} />
                ))}
            </tbody>
        </table>
    );
}

Table.propTypes = {
    headers: PropTypes.array.isRequired,
    data: PropTypes.array.isRequired,
};

export default Table;
