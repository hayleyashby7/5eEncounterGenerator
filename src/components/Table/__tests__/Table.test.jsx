import { it, describe, expect, vi, beforeEach } from 'vitest';
import React from 'react';
import { render, screen } from '@testing-library/react';
import Table from '../Table';

vi.mock('../Row');

const headers = ['Name', 'Challenge Rating', 'Type', 'Size', 'Alignment'];

describe('Table', () => {
    beforeEach(() => {
        vi.clearAllMocks();
    });

    it('should render the component', async () => {
        // Arrange
        const testData = [
            {
                name: 'Goblin',
                slug: 'goblin',
                challenge_rating: '1',
                type: 'Humanoid',
                size: 'Small',
                alignment: 'Neutral',
            },
        ];
        const mockRow = await import('../Row');
        mockRow.default.mockImplementation(() => (
            <tr>
                <td>Row</td>
            </tr>
        ));

        // Act
        const { container } = render(<Table headers={headers} data={testData} />);

        // Assert
        expect(container).toMatchSnapshot();
    });

    it('should render the correct headers', async () => {
        // Arrange
        const testData = [
            {
                name: 'Goblin',
                slug: 'goblin',
                challenge_rating: '1',
                type: 'Humanoid',
                size: 'Small',
                alignment: 'Neutral',
            },
        ];
        const mockRow = await import('../Row');
        mockRow.default.mockImplementation(() => (
            <tr>
                <td>Row</td>
            </tr>
        ));

        // Act
        render(<Table headers={headers} data={testData} />);

        // Assert
        expect(screen.getByText('Name')).toBeInTheDocument();
        expect(screen.getByText('Challenge Rating')).toBeInTheDocument();
        expect(screen.getByText('Type')).toBeInTheDocument();
        expect(screen.getByText('Size')).toBeInTheDocument();
        expect(screen.getByText('Alignment')).toBeInTheDocument();
    });

    it('should render the correct number of rows', async () => {
        // Arrange
        const testData = [
            {
                name: 'Goblin',
                slug: 'goblin',
                challenge_rating: '1',
                type: 'Humanoid',
                size: 'Small',
                alignment: 'Neutral',
            },
            {
                name: 'Orc',
                slug: 'orc',
                challenge_rating: '1/2',
                type: 'Humanoid',
                size: 'Medium',
                alignment: 'Chaotic Evil',
            },
        ];
        const mockRow = await import('../Row');
        mockRow.default.mockImplementation(() => (
            <tr>
                <td>Row</td>
            </tr>
        ));

        // Act
        render(<Table headers={headers} data={testData} />);

        // Assert
        expect(mockRow.default).toHaveBeenCalledTimes(2);
        expect(mockRow.default).toHaveBeenCalledWith(
            { data: testData[0], cols: headers.length },
            {},
        );
        expect(mockRow.default).toHaveBeenCalledWith(
            { data: testData[1], cols: headers.length },
            {},
        );
    });
});
