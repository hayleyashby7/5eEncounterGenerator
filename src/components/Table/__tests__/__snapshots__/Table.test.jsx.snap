// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > should render the component 1`] = `
<div>
  <table
    class="table w-full border-separate text-left"
  >
    <thead>
      <tr>
        <th
          class="table-cell"
        >
          Name
        </th>
        <th
          class="table-cell"
        >
          Challenge Rating
        </th>
        <th
          class="hidden md:table-cell"
        >
          Type
        </th>
        <th
          class="hidden md:table-cell"
        >
          Size
        </th>
        <th
          class="hidden md:table-cell"
        >
          Alignment
        </th>
        <th
          class="table-cell"
        />
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Row
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
